tensilelite/Tensile/Toolchain/Assembly.py

```diff
@@ -0,0 +1,226 @@
+################################################################################
+#
+# Copyright (C) 2024 Advanced Micro Devices, Inc. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+################################################################################
+
+import collections
+import math
+import os
+import shlex
+import shutil
+import subprocess
+import warnings
+
+from pathlib import Path
+from typing import List, Union, Tuple
+
+
+from .. import Utils
+from ..TensileInstructions import getGfxName, getCOVFromParam
+from ..Common import globalParameters, print2, ensurePath, printWarning, IsaVersion
+
+class AssemblyToolchain:
+    def __init__(self, assembler: str, bundler: str, buildIdKind: str):
+        self.assembler = assembler
+        self.bundler = bundler
+        self.buildIdKind = buildIdKind
+
+    def assemble(self, srcPath: str, destPath: str, coVersion: str, isa: IsaVersion, wavefrontSize: int, debug: bool=False):
+      """Assemble an assembly source file into an object file.
+
+      Args:
+          srcPath: The path to the assembly source file.
+          destPath: The destination path for the generated object file.
+          coVersion: The code object version to use.
+          isa: The target GPU architecture in ISA format.
+          wavefrontSize: The wavefront size to use.
+      """
+      coVersion = getCOVFromParam(coVersion)
+      cpu = getGfxName(isa)
+      wavefrontFlag = "-mwavefrontsize64" if wavefrontSize == 64 else "-mno-wavefrontsize64" 
+    
+      launcher = shlex.split(os.environ.get('Tensile_ASM_COMPILER_LAUNCHER', ''))
+      args = [
+          *launcher, 
+          self.assembler, 
+          "-x", "assembler", 
+          "--target=amdgcn-amd-amdhsa", 
+          f"-mcode-object-version={coVersion}", 
+          f"-mcpu={cpu}",  
+          wavefrontFlag, 
+          "-g" if debug else "",
+          "-c", 
+          "-o", destPath, srcPath
+      ]
```

Comment:
coVersion can be moved from this function into the toolchain ctor since it's specified at the program-level. The other two are more complicated since they are extracted from state on the kernelWriter. The cpu variable (which I will rename, but is actually the gfx name of an ISA) is needed to be configurable at assemble time to support multiple archs. Which would be similar for wavefront size.